pydrive2/files.py

```diff
@@ -229,17 +232,43 @@ def GetContentFile(self, filename, mimetype=None, remove_bom=False):
     :type mimetype: str
     :param remove_bom: Whether to remove the byte order marking.
     :type remove_bom: bool
+    :param callback: passed two arguments: (total trasferred, file size).
+    :type param: callable
     :raises: ApiRequestError, FileNotUploadedError, FileNotDownloadableError
     """
-        if (
-            self.content is None
-            or type(self.content) is not io.BytesIO
-            or self.has_bom == remove_bom
-        ):
-            self.FetchContent(mimetype, remove_bom)
-        f = open(filename, "wb")
-        f.write(self.content.getvalue())
-        f.close()
+        files = self.auth.service.files()
+        file_id = self.metadata.get("id") or self.get("id")
+
+        def download(fd, request):
+            downloader = MediaIoBaseDownload(fd, request)
+            done = False
+            while done is False:
+                status, done = downloader.next_chunk()
+                if callback:
+                    callback(status.resumable_progress, status.total_size)
+
+        with open(filename, mode="w+b") as fd:
+            # Ideally would use files.export_media instead if
+            # metadata.get("mimeType").startswith("application/vnd.google-apps.")
+            # but that would first require a slow call to FetchMetadata()
+            try:
+                download(fd, files.get_media(fileId=file_id))
+            except errors.HttpError as error:
+                err_str = str(error).lower()
```

Comment:
check the `ApiRequestError`: 

1. you need to wrap HttpError with it 
2. it's a proper way to test the http error code + error message/type instead of using `in`

Reply:
I was going to use `HttpError.resp.status == "403"` before you posted this. That would be similar to how `_should_retry_response` is called. I've checked the `ApiRequestError` I don't see another obvious way. Is there a `ApiRequestError().error["status"]`? Not obvious at all and also doubtful as that would mean duplication of information sent on the server side.

Reply:
Also there's this gross snippet in the unit tests:


```
    except ApiRequestError as exception:
        retry_codes = ["403", "500", "502", "503", "504"]
        if any(
            "HttpError {}".format(code) in str(exception)
            for code in retry_codes
        ):
            raise PyDriveRetriableError("Google API request failed")
```

Reply:
seriously we should fix ApiRequestError to keep the response code.

Reply:
ah got this `ApiRequestError(HttpError()).error["code"]`. Seriously? Undocumented clandestine stuff server-side.

Reply:
yep, we just preserve what Google API returns to the maximum and extract all the details on the client side as we need them - check this for some examples how to check this in a "structured" way https://github.com/iterative/dvc/blob/master/dvc/remote/gdrive.py#L47-L78

Reply:
test_util - is a bad example, let's not do it that way :)

Reply:
so, 403 can be tested. To get a reason check the whole JSON - it has reason field, location field, etc.

Reply:
I doubt they document all the reasons somewhere (even basic ones - related to rate limiting), but my understanding it is part of the protocol and at least it's more reliable than using `"str" in` check

Reply:
done